drivers/magnetometer/admt4000/admt4000.c

```diff
@@ -0,0 +1,2320 @@
+*******************************************************************************/
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "admt4000.h"
+#include "no_os_spi.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_delay.h"
+
+static uint8_t admt4000_ecc_control_registers[] = {
+	ADMT4000_02_REG_GENERAL,
+	ADMT4000_02_REG_DIGIOEN,
+	ADMT4000_02_REG_ANGLECK,
+	ADMT4000_02_REG_H1MAG,
+	ADMT4000_02_REG_H1PH,
+	ADMT4000_02_REG_H2MAG,
+	ADMT4000_02_REG_H2PH,
+	ADMT4000_02_REG_H3MAG,
+	ADMT4000_02_REG_H3PH,
+	ADMT4000_02_REG_H8MAG,
+	ADMT4000_02_REG_H8PH,
+};
+
+static uint32_t admt4000_angle_conv_factors[] = {
+	ADMT4000_ABS_ANGLE_RES,
+	ADMT4000_ANGLE_RES,
+};
+
+/***************************************************************************//**
+ * @brief Initializes the communication peripheral and checks if the admt4000
+ *        part is present.
+ *
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ *                     parameters.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int admt4000_init(struct admt4000_dev **device,
+		  struct admt4000_init_param init_param)
+{
+	bool bool_temp;
+	int ret;
+	struct admt4000_dev *dev;
+
+	if (init_param.dev_vdd < ADMT4000_3P3V || init_param.dev_vdd > ADMT4000_5V)
+		return -EINVAL;
+
+	dev = (struct admt4000_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	/* SPI */
+	ret = no_os_spi_init(&dev->spi_desc, &init_param.spi_init);
+	if (ret)
+		goto err;
+
+	/* GPIO */
+	/**  Get Descriptors **/
+	ret = no_os_gpio_get(&dev->gpio_coil_rs,
+			     &init_param.gpio_coil_rs);
+	if (ret)
+		goto err;
+
+	ret = no_os_gpio_get(&dev->gpio_gpio0_busy,
+			     &init_param.gpio_gpio0_busy);
+	if (ret)
+		goto err;
+
+	ret = no_os_gpio_get(&dev->gpio_shdn_n,
+			     &init_param.gpio_shdn_n);
+	if (ret)
+		goto err;
+
+	/** Set Direction **/
+	ret = no_os_gpio_direction_output(dev->gpio_coil_rs,
+					  NO_OS_GPIO_LOW);
+	if (ret)
+		goto err;
+
+	ret = no_os_gpio_direction_input(dev->gpio_gpio0_busy);
+	if (ret)
+		goto err;
+
+	ret = no_os_gpio_direction_output(dev->gpio_shdn_n,
+					  NO_OS_GPIO_HIGH);
+	if (ret)
+		goto err;
+
+	/* Delay for GPIO to lines to settle before clearing faults*/
+	no_os_mdelay(10);
+
+	ret = admt4000_clear_all_faults(dev);
+	if (ret)
+		goto err;
+
+	if (init_param.dev_vdd == ADMT4000_3P3V)
+		dev->fixed_conv_factor_mv = 412500; // in uV
+	else
+		dev->fixed_conv_factor_mv = 300000; // in uV
+
+	ret = admt4000_get_cnv_mode(dev, &bool_temp);
+	if (ret)
+		return ret;
+
+	dev->is_one_shot = bool_temp;
+
+	dev->is_throw_early_samples = true;
+
+	ret = admt4000_get_page(dev, &bool_temp);
+	if (ret)
+		return ret;
+
+	dev->is_page_zero = bool_temp;
+
+	*device = dev;
+
+	return 0;
+
+err:
+	admt4000_remove(dev);
```

Comment:
this would only work if you do something like this in admt4000_remove:
```
if (device->spi_desc) {
        ret = no_os_spi_remove(device->spi_desc);
	if (ret)
		return ret;
	device->spi_desc = NULL;
}
... and do this for all gpios, and whatever else was allocated ... 
```

Otherwise you need to create several goto labels like we do for other device drivers.